expand groupby() parameter

huanccwang opened the pull request:
In some business scenarios, the columns we aggregate are not fixed and unknown. We need to perform dynamic aggregation based on business configuration information, so we need to pass in dynamic columns,I looked at the source code and found that when multiple columns are passed in, it is also converted to Seq, so I expanded the parameters of groupBy() to support passing in Seq.

Commit:
expand groupBy parameter

Commit:
Merge branch 'master' into spark_groupBy_parameter_expand

juliuszsompolski committed:
[SPARK-46075][CONNECT] Improvements to SparkConnectSessionManager

### What changes were proposed in this pull request?

This is factored out from https://github.com/apache/spark/pull/43913 and is a continuation to https://github.com/apache/spark/pull/43546 when SparkConnectSessionManager was introduced.

We want to remove the use a Guava cache as session cache, and have our custom logic with more control. This refactors the Session Manager and adds more tests.

We introduce a mechanism that mirrors SparkConnectExecutionManager instead.

### Why are the changes needed?

With guava cache, only a single "inactivity timeout" can be specified for the whole cache. This can't be for example overriden per session. The actual invalidation also happens not in it's own thread inside guava, but it's work-stealing lazily piggy backed to other operations on the cache, making it opaque when session removal will actually happen.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

SparkConnectSessionManagerSuite added.

### Was this patch authored or co-authored using generative AI tooling?

Github Copilot was assisting in some boilerplate auto-completion.

Generated-by: Github Copilot

Closes #43985 from juliuszsompolski/SPARK-46075.

Authored-by: Juliusz Sompolski <julek@databricks.com>
Signed-off-by: Hyukjin Kwon <gurwls223@apache.org>

dongjoon-hyun committed:
[SPARK-46353][CORE] Refactor to improve `RegisterWorker` unit test coverage

### What changes were proposed in this pull request?

This PR aims to improve the unit test coverage for `RegisterWorker` message handling.

- Add `handleRegisterWorker` helper method which is testable easily.
- Add new unit tests for three conditional branches.

### Why are the changes needed?

It's easily to test and improve. We can add more tests in this way in the future.

### Does this PR introduce _any_ user-facing change?

No. This is a refactoring on the main code and only additions to the test methods.

### How was this patch tested?

Pass the CIs.

### Was this patch authored or co-authored using generative AI tooling?

No.

Closes #44284 from dongjoon-hyun/SPARK-46353.

Authored-by: Dongjoon Hyun <dhyun@apple.com>
Signed-off-by: Dongjoon Hyun <dhyun@apple.com>

cloud-fan committed:
[SPARK-46378][SQL] Still remove Sort after converting Aggregate to Project

### What changes were proposed in this pull request?

This is a follow-up of https://github.com/apache/spark/pull/33397 to avoid sub-optimal plans. After converting `Aggregate` to `Project`, there is information lost: `Aggregate` doesn't care about the data order of inputs, but `Project` cares. `EliminateSorts` can remove `Sort` below `Aggregate`, but it doesn't work anymore if we convert `Aggregate` to `Project`.

This PR fixes this issue by tagging the `Project` to be order-irrelevant if it's converted from `Aggregate`. Then `EliminateSorts` optimizes the tagged `Project`.

### Why are the changes needed?

avoid sub-optimal plans

### Does this PR introduce _any_ user-facing change?

No
### How was this patch tested?

new test

### Was this patch authored or co-authored using generative AI tooling?

No

Closes #44310 from cloud-fan/sort.

Authored-by: Wenchen Fan <wenchen@databricks.com>
Signed-off-by: Wenchen Fan <wenchen@databricks.com>

allisonwang-db committed:
[SPARK-46253][PYTHON] Plan Python data source read using MapInArrow

### What changes were proposed in this pull request?

This PR changes how we plan Python data source read. Instead of using a regular Python UDTF, we can use an arrow UDF and plan the data source read using the MapInArrow operator.

### Why are the changes needed?

To improve the performance

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Unit tests

### Was this patch authored or co-authored using generative AI tooling?

No

Closes #44170 from allisonwang-db/spark-46253-arrow-read.

Authored-by: allisonwang-db <allison.wang@databricks.com>
Signed-off-by: Hyukjin Kwon <gurwls223@apache.org>

vicennial committed:
[SPARK-46202][CONNECT] Expose new ArtifactManager APIs to support custom target directories

### What changes were proposed in this pull request?
Adds new client APIs to the Spark Connect Scala Client:
- `def addArtifact(bytes: Array[Byte], target: String): Unit`
- `def addArtifact(source: String, target: String): Unit`

### Why are the changes needed?
Currently, without the use of a REPL/Class finder, there is no API to support adding artifacts (file-based and in-memory)
 with a custom target directory structure to the remote Spark Connect session.

### Does this PR introduce _any_ user-facing change?
Yes.

Users can do the following for classfiles and jars:
```scala
addArtifact("/Users/dummyUser/files/foo/bar.class", "sub/directory/foo.class")
addArtifact(bytesBar, "bar.class")
```

This would preserve the directory structure in the remote server. In this case, the file would be stored under the directory:
`$ROOT_SESSION_ARTIFACT_DIRECTORY/classes/bar.class`
`$ROOT_SESSION_ARTIFACT_DIRECTORY/classes/sub/directory/foo.class`
### How was this patch tested?

Unit test

### Was this patch authored or co-authored using generative AI tooling?

No.

Closes #44109 from vicennial/SPARK-46202.

Authored-by: vicennial <venkata.gudesa@databricks.com>
Signed-off-by: Herman van Hovell <herman@databricks.com>

dtenedor committed:
[SPARK-46370][SQL] Fix bug when querying from table after changing column defaults

### What changes were proposed in this pull request?

This PR fixes a bug when querying from table after changing defaults:

```
drop table if exists t;
create table t(i int, s string default 'def') using parquet;
insert into t select 1, default;
alter table t alter column s drop default;
insert into t select 2, default;
select * from t;  -- Removing this line changes the following results!
alter table t alter column s set default 'mno';
insert into t select 3, default;
select * from t;
```

The bug is related to the relation cache, and the fix involves adding a manual refresh to the cache to make sure we use the right table schema.

### Why are the changes needed?

This PR fixes a correctness bug.

### Does this PR introduce _any_ user-facing change?

Yes, see above.

### How was this patch tested?

This PR adds test coverage.

### Was this patch authored or co-authored using generative AI tooling?

No.

Closes #44302 from dtenedor/fix-default-bug.

Authored-by: Daniel Tenedorio <daniel.tenedorio@databricks.com>
Signed-off-by: Wenchen Fan <wenchen@databricks.com>

hasnain-db committed:
[SPARK-46132][CORE] Support key password for JKS keys for RPC SSL

### What changes were proposed in this pull request?

Add support for a separate key password in addition to the key store password for JKS keys. This is needed for keys which may have a key password in addition to a key store password. We already had this support for the UI SSL support, so for compatibility we should have it here.

This wasn't done earlier as I wasn't sure how to implement it but the discussion in https://github.com/apache/spark/pull/43998#discussion_r1406993411 suggested the right way.

### Why are the changes needed?

These are needed to support users who may have such keys configured.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Added some unit tests

```
build/sbt
> project network-common
> testOnly
```

### Was this patch authored or co-authored using generative AI tooling?

No

Closes #44264 from hasnain-db/separate-pw.

Authored-by: Hasnain Lakhani <hasnain.lakhani@databricks.com>
Signed-off-by: Mridul Muralidharan <mridul<at>gmail.com>

bersprockets commented:
Your IDE seems to have reformatted Dataset.scala such that several hundred lines appear changed. It would be difficult to tell what your change is. Also, your Dataset.scala no longer follows Spark coding style. See https://spark.apache.org/contributing.html "Code style guide".

See https://spark.apache.org/contributing.html, "Pull request" for guidelines on making the pull request (and also on how to get the test workflow to run on your PR).

huanccwang commented:
> Your IDE seems to have reformatted Dataset.scala such that several hundred lines appear changed. It would be difficult to tell what your change is. Also, your Dataset.scala no longer follows Spark coding style. See https://spark.apache.org/contributing.html "Code style guide".
> 
> Edit: Probably best to re-apply your change to Dataset.scala without the reformatting.
> 
> See https://spark.apache.org/contributing.html, "Pull request" for guidelines on making the pull request (and also on how to get the test workflow to run on your PR).

ok

zhengruifeng committed:
[SPARK-46379][PS][TESTS] Reorganize `FrameInterpolateTests`

### What changes were proposed in this pull request?
Reorganize `FrameInterpolateTests`:
1, move it to the right place;
2, optimize it by control the param combinations;
3, make it more stable by add `sort_index`;

### Why are the changes needed?
clean up test code

### Does this PR introduce _any_ user-facing change?
no, test-only

### How was this patch tested?
ci

### Was this patch authored or co-authored using generative AI tooling?
no

Closes #44313 from zhengruifeng/ps_test_df_interpolate.

Authored-by: Ruifeng Zheng <ruifengz@apache.org>
Signed-off-by: Hyukjin Kwon <gurwls223@apache.org>

MaxGekk committed:
[SPARK-46381][SQL] Migrate sub-classes of `AnalysisException` to error classes

### What changes were proposed in this pull request?
In the PR, I propose to migrate the rest two sub-classes of `AnalysisException` onto error classes:
- NonEmptyNamespaceException
- ExtendedAnalysisException

and forbid raising of such exception without an error class.

### Why are the changes needed?
This is a part of the migration on the error framework, and unifying errors in Spark.

### Does this PR introduce _any_ user-facing change?
Yes, if user's code depends on the format of error messages.

### How was this patch tested?
By existing test suites like:
```
$ PYSPARK_PYTHON=python3 build/sbt "sql/testOnly org.apache.spark.sql.SQLQueryTestSuite"
```

### Was this patch authored or co-authored using generative AI tooling?
No.

Closes #44314 from MaxGekk/error-class-ExtendedAnalysisException.

Authored-by: Max Gekk <max.gekk@gmail.com>
Signed-off-by: Max Gekk <max.gekk@gmail.com>

yaooqinn committed:
[SPARK-46388][SQL] HiveAnalysis misses the pattern guard of `query.resolved`

### What changes were proposed in this pull request?

This PR adds `query.resolved` as a pattern guard when HiveAnalysis converts InsertIntoStatement to InsertIntoHiveTable.

### Why are the changes needed?

Due to https://github.com/apache/spark/pull/41262/files#diff-ed19f376a63eba52eea59ca71f3355d4495fad4fad4db9a3324aade0d4986a47R1080, the `table` field is resolved regardless of the query field. Before, it never got a chance to be resolved as `HiveTableRelation` and then match any rule of HiveAnalysis. But now, it gets the chance always and results in a spark-kernel bug - `Invalid call to toAttribute on unresolved object.`

```
insert into t2 select cast(a as short) from t where b=1;
Invalid call to toAttribute on unresolved object
```

### Does this PR introduce _any_ user-facing change?

no, bugfix for 3.5 and later

### How was this patch tested?

added new test

### Was this patch authored or co-authored using generative AI tooling?

no

Closes #44326 from yaooqinn/SPARK-46388.

Authored-by: Kent Yao <yao@apache.org>
Signed-off-by: Kent Yao <yao@apache.org>

sandip-db committed:
[SPARK-46153][SQL] XML: Add TimestampNTZType support

### What changes were proposed in this pull request?
Add TimestampNTZType support in XML data source.

### Why are the changes needed?
To bring in parity with json/csv.

### Does this PR introduce _any_ user-facing change?
Yes, adds support for TimestampNTZType.

### How was this patch tested?
New unit tests.

### Was this patch authored or co-authored using generative AI tooling?
No

Closes #44329 from sandip-db/xml-timestampNTZ.

Authored-by: Sandip Agarwala <131817656+sandip-db@users.noreply.github.com>
Signed-off-by: Hyukjin Kwon <gurwls223@apache.org>

juliuszsompolski committed:
[SPARK-45798][CONNECT] Followup: add serverSessionId to SessionHolderInfo

### What changes were proposed in this pull request?

Small followup to https://github.com/apache/spark/pull/43664 - add serverSessionId to SessionHolderInfo.

### Why are the changes needed?

SessionHolderInfo should contain this kind of information about the session.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

NA

### Was this patch authored or co-authored using generative AI tooling?

No.

Closes #44334 from juliuszsompolski/SPARK-45798-fup.

Authored-by: Juliusz Sompolski <julek@databricks.com>
Signed-off-by: Hyukjin Kwon <gurwls223@apache.org>

itholic committed:
[SPARK-46387][PS][DOCS] Add an information of `isocalendar` into migration guide

### What changes were proposed in this pull request?

This PR proposes to add an information of `isocalendar` into migration guide

### Why are the changes needed?

We introduce `isocalendar` from https://github.com/apache/spark/pull/40420, but didn't notice in anywhere from our documentation.

### Does this PR introduce _any_ user-facing change?

No, but it affects the user-facing documentation.

### How was this patch tested?

The existing CI should pass.

### Was this patch authored or co-authored using generative AI tooling?

No.

Closes #44325 from itholic/migration_guide_followup.

Authored-by: Haejoon Lee <haejoon.lee@databricks.com>
Signed-off-by: Hyukjin Kwon <gurwls223@apache.org>

xinrong-meng committed:
[SPARK-46385][PYTHON][TESTS] Test aggregate functions for groups (pyspark.sql.group)

### What changes were proposed in this pull request?
Test aggregate functions for groups (pyspark.sql.group)

### Why are the changes needed?
Subtasks of [SPARK-46041](https://issues.apache.org/jira/browse/SPARK-46041) to improve test coverage

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Test change only.

### Was this patch authored or co-authored using generative AI tooling?
No.

Closes #44322 from xinrong-meng/test_group.

Authored-by: Xinrong Meng <xinrong@apache.org>
Signed-off-by: Hyukjin Kwon <gurwls223@apache.org>

itholic committed:
[SPARK-46360][PYTHON][FOLLOWUP][DOCS] Add `getMessage` to API reference

### What changes were proposed in this pull request?

This PR followups https://github.com/apache/spark/pull/44292 to add documentation.

### Why are the changes needed?

To address missing documentation.

### Does this PR introduce _any_ user-facing change?

No API changes, but `getMessage` is added to API reference page.

### How was this patch tested?

The existing doc build in CI should pass.

### Was this patch authored or co-authored using generative AI tooling?

No.

Closes #44308 from itholic/getMessage_followup.

Authored-by: Haejoon Lee <haejoon.lee@databricks.com>
Signed-off-by: Hyukjin Kwon <gurwls223@apache.org>

nchammas committed:
[SPARK-46357] Replace incorrect documentation use of setConf with conf.set

### What changes were proposed in this pull request?

`setConf` is a method on `SQLContext`. The docs perhaps used to refer to `SQLContext`, but now they refer to `SparkSession`, which does not have this method.

This PR updates the docs to use the appropriate method for making configuration settings against `SparkSession`.

### Why are the changes needed?

The current documentation is incorrect.

### Does this PR introduce _any_ user-facing change?

Yes.

### How was this patch tested?

No testing.

### Was this patch authored or co-authored using generative AI tooling?

No.

Closes #44290 from nchammas/SPARK-46357-setConf-confset.

Authored-by: Nicholas Chammas <nicholas.chammas@gmail.com>
Signed-off-by: Kent Yao <yao@apache.org>

panbingkun committed:
[SPARK-46377][INFRA] Upgrade labeler action to v5

### What changes were proposed in this pull request?
The pr aims to upgrade `labeler action` from v4 to v5.

### Why are the changes needed?
The full release notes: https://github.com/actions/labeler/releases/tag/v5.0.0
Version 5 of this action updated the [runtime to Node.js 20](https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#runs-for-javascript-actions). All scripts are now run with Node.js 20 instead of Node.js 16 and are affected by any breaking changes between Node.js 16 and 20.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
- Pass GA.
- Manually continuous observation.

### Was this patch authored or co-authored using generative AI tooling?
No.

Closes #44309 from panbingkun/SPARK-46377.

Authored-by: panbingkun <pbk1982@gmail.com>
Signed-off-by: Kent Yao <yao@apache.org>

panbingkun committed:
Revert "[SPARK-46377][INFRA] Upgrade labeler action to v5"

### What changes were proposed in this pull request?
This reverts commit 270da6f9b7b331e455d1f0bd1309fb87bc8740ab.

### Why are the changes needed?
Master CI failed:

```
Run actions/labelerv5
The configuration file (path: .github/labeler.yml) was not found locally, fetching via the api
Error: Error: found unexpected type for label 'INFRA' (should be array of config options)
Error: found unexpected type for label 'INFRA' (should be array of config options)
```

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Pass GA

### Was this patch authored or co-authored using generative AI tooling?
No

Closes #44344 from panbingkun/SPARK-46377_revert.

Authored-by: panbingkun <pbk1982@gmail.com>
Signed-off-by: yangjie01 <yangjie01@baidu.com>

anchovYu committed:
[SPARK-46394][SQL] Fix spark.catalog.listDatabases() issues on schemas with special characters when `spark.sql.legacy.keepCommandOutputSchema` set to true

### What changes were proposed in this pull request?
When the SQL conf `spark.sql.legacy.keepCommandOutputSchema` is set to true:
Before:
```
// support there is a xyyu-db-with-hyphen schema in the catalog
spark.catalog.listDatabases()

[INVALID_IDENTIFIER] The identifier xyyu-db-with-hyphen is invalid. Please, consider quoting it with back-quotes as `xyyu-db-with-hyphen`. SQLSTATE: 42602 (line 1, pos 4)
```

After:
```
spark.catalog.listDatabases()

.. `xyyu-db-with-hyphen` ..
```

This PR fixes the issue by falling back to original name when the parsing failed.

### Why are the changes needed?
To fix the bug.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Newly added tests.

### Was this patch authored or co-authored using generative AI tooling?
No.

Closes #44328 from anchovYu/special-char-schema-name-issue.

Authored-by: Xinyi Yu <xinyi.yu@databricks.com>
Signed-off-by: Wenchen Fan <wenchen@databricks.com>

zhengruifeng committed:
[SPARK-46379][PS][TESTS][FOLLOWUPS] Deduplicate `test_interpolate_error`

### What changes were proposed in this pull request?
this is a followup of https://github.com/apache/spark/pull/44313/, which happened to duplicate `test_interpolate_error`

### Why are the changes needed?
don't need to test it twice

### Does this PR introduce _any_ user-facing change?
no

### How was this patch tested?
ci

### Was this patch authored or co-authored using generative AI tooling?
no

Closes #44341 from zhengruifeng/frame_interpolate_followup.

Authored-by: Ruifeng Zheng <ruifengz@apache.org>
Signed-off-by: Ruifeng Zheng <ruifengz@apache.org>

MaxGekk committed:
[SPARK-46393][SQL] Classify exceptions in the JDBC table catalog

### What changes were proposed in this pull request?
In the PR, I propose to handle exceptions from JDBC drivers in the JDBC table catalog, classify them and converted to appropriate Spark exception w/ an error class. This PR covers the following functions where such errors haven't been classified yet:
- list tables
- namespace exists
- list namespaces

### Why are the changes needed?
To unify Spark exceptions, and migrate onto new error framework.

### Does this PR introduce _any_ user-facing change?
Yes, if user code expects that Spark SQL bypass Java exceptions from JDBC drivers.

### How was this patch tested?
By existing test suites:
```
$ build/sbt "test:testOnly *JDBCV2Suite"
$ build/sbt "test:testOnly *JDBCTableCatalogSuite"
```

### Was this patch authored or co-authored using generative AI tooling?
No.

Closes #44335 from MaxGekk/classifyException-JDBCTableCatalog.

Authored-by: Max Gekk <max.gekk@gmail.com>
Signed-off-by: Max Gekk <max.gekk@gmail.com>

gengliangwang committed:
[SPARK-46396][SQL] Timestamp inference should not throw exception

### What changes were proposed in this pull request?

When setting `spark.sql.legacy.timeParserPolicy=LEGACY`, Spark will use the LegacyFastTimestampFormatter to infer potential timestamp columns. The inference shouldn't throw exception.

However, when the input is 23012150952, there is exception:

```

For input string: "23012150952"

java.lang.NumberFormatException: For input string: "23012150952"

at java.base/java.lang.NumberFormatException.forInputString(NumberFormatException.java:67)

at java.base/java.lang.Integer.parseInt(Integer.java:668)

at java.base/java.lang.Integer.parseInt(Integer.java:786)

at org.apache.commons.lang3.time.FastDateParser$NumberStrategy.parse(FastDateParser.java:304)

at org.apache.commons.lang3.time.FastDateParser.parse(FastDateParser.java:1045)

at org.apache.commons.lang3.time.FastDateFormat.parse(FastDateFormat.java:651)

at org.apache.spark.sql.catalyst.util.LegacyFastTimestampFormatter.parseOptional(TimestampFormatter.scala:418)

```

This PR is to fix the issue.

### Why are the changes needed?

Bug fix, Timestamp inference should not throw exception
### Does this PR introduce _any_ user-facing change?

NO

### How was this patch tested?

New test case + existing tests

### Was this patch authored or co-authored using generative AI tooling?

No

Closes #44338 from gengliangwang/fixParseOptional.

Authored-by: Gengliang Wang <gengliang@apache.org>
Signed-off-by: Gengliang Wang <gengliang@apache.org>

LuciferYang committed:
[SPARK-46401][CORE] Use `!isEmpty()` on `RoaringBitmap` instead of `getCardinality() > 0` in `RemoteBlockPushResolver`

### What changes were proposed in this pull request?
This pr use `!mapTracker.isEmpty()` instead of `mapTracker.getCardinality() > 0` in `RemoteBlockPushResolver ` to reduce some computational costs, as the `getCardinality()` method triggers a recount every time it is called.

- [getCardinality](https://github.com/RoaringBitmap/RoaringBitmap/blame/578a5162fa7c80be3988bfdbf9abe06ae124722c/RoaringBitmap/src/main/java/org/roaringbitmap/RoaringBitmap.java#L1884-L1901)

```java
  /**
   * Returns the number of distinct integers added to the bitmap (e.g., number of bits set).
   *
   * return the cardinality
   */
  Override
  public long getLongCardinality() {
    long size = 0;
    for (int i = 0; i < this.highLowContainer.size(); i++) {
      size += this.highLowContainer.getContainerAtIndex(i).getCardinality();
    }
    return size;
  }

  Override
  public int getCardinality() {
    return (int) getLongCardinality();
  }
```

- [isEmpty](https://github.com/RoaringBitmap/RoaringBitmap/blob/578a5162fa7c80be3988bfdbf9abe06ae124722c/RoaringBitmap/src/main/java/org/roaringbitmap/RoaringBitmap.java#L2218-L2226)

```java
 /**
   * Checks whether the bitmap is empty.
   *
   * return true if this bitmap contains no set bit
   */
  Override
  public boolean isEmpty() {
    return highLowContainer.size() == 0;
  }
```

This change is refer to https://github.com/RoaringBitmap/RoaringBitmap/issues/239 | https://github.com/RoaringBitmap/RoaringBitmap/pull/684

### Why are the changes needed?
Use a more appropriate API to reduce the computational cost of empty-checking for RoaringBitmap.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Pass GitHub Actions

### Was this patch authored or co-authored using generative AI tooling?
No

Closes #44347 from LuciferYang/bitmap-isEmpty.

Authored-by: yangjie01 <yangjie01@baidu.com>
Signed-off-by: yangjie01 <yangjie01@baidu.com>

zhengruifeng committed:
[SPARK-46391][PS][TESTS] Reorganize `ExpandingParityTests`

### What changes were proposed in this pull request?
Reorganize `ExpandingParityTests`

### Why are the changes needed?
to make the test more consistent with pandas

### Does this PR introduce _any_ user-facing change?
no, test-only

### How was this patch tested?
ci

### Was this patch authored or co-authored using generative AI tooling?
no

Closes #44332 from zhengruifeng/ps_test_expanding.

Authored-by: Ruifeng Zheng <ruifengz@apache.org>
Signed-off-by: Ruifeng Zheng <ruifengz@apache.org>

beliefer committed:
[SPARK-45796][SQL] Support MODE() WITHIN GROUP (ORDER BY col)

### What changes were proposed in this pull request?
The mainstream database supports this syntax.
**H2**
http://www.h2database.com/html/functions-aggregate.html#mode
**Postgres**
https://www.postgresql.org/docs/16/functions-aggregate.html

**Syntax**:
Aggregate function
`MODE() WITHIN GROUP (ORDER BY sortSpecification)`

Window function
```
MODE() WITHIN GROUP (ORDER BY sortSpecification)
[FILTER (WHERE expression)] [OVER windowNameOrSpecification]
```

**Examples**:
```
SELECT
  mode() WITHIN GROUP (ORDER BY v),
  mode() WITHIN GROUP (ORDER BY v) FILTER (WHERE k > 0)
FROM aggr;
```

```
SELECT
    employee_name,
    department,
    salary,
    mode() WITHIN GROUP (ORDER BY salary) OVER (PARTITION BY department),
    mode() WITHIN GROUP (ORDER BY salary) FILTER (WHERE department = 'Accounting') OVER (PARTITION BY department)
FROM basic_pays
ORDER BY salary;
```

### Why are the changes needed?
Support `MODE() WITHIN GROUP (ORDER BY col)`

### Does this PR introduce _any_ user-facing change?
'No'.
New feature.

### How was this patch tested?
New test cases.

### Was this patch authored or co-authored using generative AI tooling?
'No'.

Closes #44184 from beliefer/SPARK-45796_new.

Authored-by: Jiaan Geng <beliefer@163.com>
Signed-off-by: Jiaan Geng <beliefer@163.com>

Commit:
format

Commit:
Merge branch 'temp' into spark_groupBy_parameter_expand

# Conflicts:
#	sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala
